server/resp/src/main/java/org/infinispan/server/resp/commands/json/JSONSET.java

```diff
@@ -0,0 +1,139 @@
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.jayway.jsonpath.Configuration;
+import com.jayway.jsonpath.JsonPath;
+import com.jayway.jsonpath.Option;
+import com.jayway.jsonpath.PathNotFoundException;
+
+import io.netty.channel.ChannelHandlerContext;
+
+/**
+ * JSON.SET
+ *
+ * @see <a href="https://redis.io/commands/json.set/">JSON.SET</a>
+ * @since 15.1
+ */
+public class JSONSET extends RespCommand implements Resp3Command {
+
+   public static final String XX = "XX";
+   public static final String NX = "NX";
+   // Configure JsonPath to use Jackson. Path with missing final leaf will return null
+   // path with more the 1 level missing will rise exception
+   static Configuration config = Configuration.builder().options(Option.DEFAULT_PATH_LEAF_TO_NULL)
+         .jsonProvider(new com.jayway.jsonpath.spi.json.JacksonJsonNodeJsonProvider())
+         .mappingProvider(new com.jayway.jsonpath.spi.mapper.JacksonMappingProvider())
+         .build();
+
+   public static void jsonSetBiConsumer(CharSequence value, ResponseWriter writer) {
+      if (value == null) {
+         // Use null consumer
+         writer.nulls();
+         return;
+      }
+      if (value.length() == 2 && value.charAt(0) == 'O' && value.charAt(1) == 'K') {
+         // Use ok consumer
+         writer.ok();
+         return;
+      }
+      // All the rest are errors
+      writer.error(value);
+   }
+
+   public JSONSET() {
+      super("JSON.SET", -3, 1, 1, 1);
+   }
+
+   @Override
+   public CompletionStage<RespRequestHandler> perform(Resp3Handler handler, ChannelHandlerContext ctx,
+         List<byte[]> arguments) {
+      byte[] key = arguments.get(0);
+      byte[] path = arguments.get(1);
+      byte[] value = arguments.get(2);
+      if (arguments.size() > 4) {
+         handler.writer().syntaxError();
+         return handler.myStage();
+      }
+      boolean nx, xx; // must be effectively final
+      if (arguments.size() == 4) {
+         String arg = (new String(arguments.get(3)).toUpperCase());
+         switch (arg) {
+            case NX:
+               nx = true;
+               xx = false;
+               break;
+            case XX:
+               xx = true;
+               nx = false;
+               break;
+            default:
+               handler.writer().syntaxError();
+               return handler.myStage();
+         }
+      } else {
+         nx = false;
+         xx = false;
+      }
+      ObjectMapper mapper = new ObjectMapper();
+         JsonNode newNode;
+         try {
+            newNode = mapper.readTree(value);
+         } catch (IOException e) {
+            handler.writer().error(e.getMessage());
+            return handler.myStage();
+         }
+         FunctionalMap.ReadWriteMap<byte[], Object> cache = ReadWriteMapImpl
+               .create(FunctionalMapImpl.create(handler.typedCache(null)));
+         CompletionStage<CharSequence> cs = cache.eval(key, view -> {
```

Comment:
Wondering if we should utilize the base `compute` method in the cache. If we go with the functional map, we'll likely want to create an instance with the other caches and return with `handler.functionalMap()` or something like that. Also, I am not 100% sure about the behavior during a transaction, since we create a new instance :(

Reply:
going to move to cache.computeAsync

Reply:
mmm, not that simple. 
cache.compute returns the new value and we don't need it. giving that json doc could be large, maybe we don't want this